### PR TITLE
Add ES6 Symbol polyfill for compatibility with iOS 8 and Android

### LINKED ISSUE #4676: Polyfill ES6 Symbol
Hi, my env:
react-native-cli : 0.1.7
react-native : 0.16.0
[react-native-icons](https://github.com/corymsmith/react-native-icons): 0.7.0

when I run my app ,I got the following error:
```
E/ReactNativeJS( 2973): Can't find variable: Symbol
```

but the Example of react-native-icons works fine.

### PR DESCRIPTION
**Pull request by skevy**
Fixes #4676

### CODE REVIEW

**Comment by martinbigio** [comment]
cc @davidaurelio

**Comment by mkonicek** [comment]
Thanks a lot for the fix @skevy!

@davidaurelio How do we merge PRs that add dependencies now that we have the new infra for storing node_modules internally?

For context, we have a custom way to fetch node_modules at fb because React Native is used in the main fb apps and there are engineers working on those apps who don't deal with JS, npm or anything like that. node_modules are kept up to date transparently. We used to check node_modules into the Mercurial repo but maintaining that proved to be a pain so we have a new system to do that.

**Comment by davidaurelio** [comment]
@mkonicek I guess for now we have to have these imported by a FB employee and do the necessary things. We could enhance the bot to a) require that or b) do the necessary things itself.

File: Libraries/JavaScriptAppEngine/Initialization/InitializeJavaScriptAppEngine.js [hunk 1]
```diff
@@ -126,6 +126,10 @@ function setUpAlert() {
   }
 }
 
+function setUpSymbol() {
+  GLOBAL.Symbol = require('es6-symbol');
```
Comment (davidaurelio) [inline Libraries/JavaScriptAppEngine/Initialization/InitializeJavaScriptAppEngine.js]: I don’t think we want to overwrite native `Symbol` if present.

Comment (davidaurelio) [inline Libraries/JavaScriptAppEngine/Initialization/InitializeJavaScriptAppEngine.js]: i.e. use `require('es6-symbol/implement');`

Comment (skevy) [inline Libraries/JavaScriptAppEngine/Initialization/InitializeJavaScriptAppEngine.js]: https://github.com/medikoo/es6-symbol/blob/master/index.js#L3

Am I reading that wrong? It seems like it's checking for native, global `Symbol` before it polyfills.

Or would you rather have that check in our code?

Comment (davidaurelio) [inline Libraries/JavaScriptAppEngine/Initialization/InitializeJavaScriptAppEngine.js]: I’m sorry, I misunderstood their documentation.

Comment (skevy) [inline Libraries/JavaScriptAppEngine/Initialization/InitializeJavaScriptAppEngine.js]: @davidaurelio np :) If you'd rather me move the check into RN so it's more clear, I'm happy to do that!

Comment (davidaurelio) [inline Libraries/JavaScriptAppEngine/Initialization/InitializeJavaScriptAppEngine.js]: no, it’s good the way it is. @mkonicek will take care of merging

Comment (skevy) [inline Libraries/JavaScriptAppEngine/Initialization/InitializeJavaScriptAppEngine.js]: Thanks gentleman!

**Comment by mkonicek** [comment]
@facebook-github-bot import

**Comment by mkonicek** [comment]
I just did `npm install --save symbol@3.0.2` which added 973 JS files to node_modules, making React Native JS deps more bloated and making the JS bundle bigger. Maybe this is how the JS ecosystem works but adding so much code to polyfill `Symbol` worth it? :(

Here are all the file: http://pastebin.com/wyePxHtd

**Comment by ide** [comment]
Could we use core-js instead of es6-symbol, since that's what Babel uses (or used to use...) and should get deduped with Babel?

**Comment by satya164** [comment]
What about using `babel-polyfill`?

**Comment by skevy** [comment]
@mkonicek sorry I didn't see this. I'll take a look.

I'm sure we can figure out a different solution.

We actually do already have Babel-polyfill in deps (not actually sure why)...I was just hesitant of using that as it adds a lot of code to people's bundles and doesn't really add that much (since we polyfill a bunch anyway).

I'm AFK at the moment, but my intuition tells me that either we can vendor a symbol polyfill or try core-JS...

**Comment by skevy** [comment]
Honestly, I wouldn't be surprised if the deps that es6-symbol added are deps we already have somewhere in the tree...but that's not how npm works. :-(

**Comment by mkonicek** [comment]
@skevy Yes I was just worried when I saw this was adding files like `node_modules/es6-symbol/node_modules/es5-ext/math/asinh/implement.js` that are very likely not needed.

I have little context on what Symbol is. If you guys more familiar with JS are happy with merging this go ahead. I was just asking whether adding all this code was worth it, and you're right all of it might be already in the tree :(

It's frustrating to see that the `node_modules` is 294MB (!) for a minimal RN app (with npm2), and I understand that npm3 is even slower than npm2 so we can't recommend it yet.

**Comment by skevy** [comment]
@mkonicek totally agreed. Hold off on merging this everyone - I'm gonna just vendorize this polyfill -- absolutely no reason to add a huge dep.

**Comment by skevy** [comment]
@mkonicek I am going to do this right now though, so hopefully we can still get this into RN 0.18...would be a good fix as this causes a problem on < iOS 9 and Android (some or all...not sure).

**Comment by skevy** [comment]
OK so actually, two comments on this:

1) Like @ide says, we could use core-js here. I don't think including core-js as a dependency is a good solution though, as that would still contribute to node-module bloat (at least on npm2...npm3 would de-dupe if we use the same semver range as babel). But, we could do a custom build of core-js with the Symbol polyfill only. I just did that...it's 24KB. Problem is, the code is pretty unreadable (it's webpack generated). I usually wouldn't mind this, since it's just a polyfill and no one is liable to touch it...but all the other polyfills that currently are vendorized in RN are extremely readable.

2) We're already currently including `babel-polyfill` as a dependency. I've used `babel-polyfill` in my projects, and it's caused no issues. It would add some bloat to the actual bundle size...I can see how much if you guys think it would be reasonable to explore this option. @satya164

**Comment by ide** [comment]
The size of the bundle is more important than the size of node_modules here. npm uses gzip and npm 3 is not bandwidth-constrained, and 3xxMB is a fraction of a percent of modern HDD space. So whatever polyfill gives us some combination of compact and clean code is actually my preference, whether that's core-js or something else.

**Comment by skevy** [comment]
I'm talking to @vjeux about this...will report back.

**Comment by vjeux** [comment]
Here's the plan:

1) Import the `babel-plugin-symbol-member` transform that we have [internally](https://phabricator.fb.com/diffusion/E/browse/tfb/trunk/www/scripts/static_resources/js/fb-transforms/babel-6/babel-plugin-symbol-member/): https://gist.github.com/vjeux/2715c9687653547037b0 

It just replaces `Symbol.iterator` with `typeof Symbol.iterator === 'function' ? Symbol.iterator : '@@iterator'` which makes for..of work. No need to include a 23kb polyfill since the only use case is to have this variable set.

2) Update the JavaScript environment docs to say that we support for..of and mention the existence of this polyfill. http://facebook.github.io/react-native/docs/javascript-environment.html#content

**Comment by mkonicek** [comment]
Nice! I've released 0.18.0. Can anyone release 0.18.1 with the single extra commit? Since it's a JS-only change, it's just `npm publish`.

**Comment by satya164** [comment]
ping @bestander :)

**Comment by vjeux** [comment]
We don't want to merge this pull request, see the plan outlined above.

**Comment by skevy** [comment]
@vjeux I did a bit more investigation into our plan.

The transform works, but it doesn't REALLY solve the use case. In fact, neither did this original PR (using es6-symbol). The MOST common use case of Symbol, in modern JS code, is with for..of. For `for..of` to work, it requires two things:

1) `Symbol.iterator` needs to exist (solved by `es6-symbol` or more simply, `babel-plugin-symbol-member`).
2) `Array.prototype[@@iterator]()` needs to exist (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/@@iterator). This is obviously a bit tougher, and involves actually writing some code to shim the iterator functionality.

I don't think this is going to get solved this way.

My recommended course of action is that we update the docs to say that:

1) In RN core, while we polyfill certain modules, we choose to be careful about what we polyfill and what we don't (as polyfills often introduce performance concerns), and as such, we don't include a heavy-duty, one-size-fits-all polyfill (such as core-js) by default. This means that, given the differences between JSC in iOS 9, iOS 8, and the version we include on Android (which is pretty old), some ES2015 features are not fully supported. These include (but may not be limited to):

* Iterators + For..Of
* ~~Generators~~ (we support these through regenerator)
* ~~Map~~ (supported through included custom polyfill)
* ~~Set~~ (supported through included custom polyfill)
* WeakMap
* WeakSet
* Proxy
* Symbols (we could actually still polyfill this through the `symbol-member` plugin, but I'd recommend against that to avoid confusion)
* Certain Math + Number + String + Object APIs may not be supported across all platforms.
* ~~Promises~~ (extra functionality supported through our polyfill)
* Reflect API

2) If your application wants or needs any of the features that are not supported out-of-the-box with RN, we can recommend investigating using babel-polyfill, core-js, etc. But should someone choose to use one of these and have a problem, this is unsupported behavior and they do so at their own risk.

3) Reiterate that RN developers SHOULD NOT rely on debugging in Chrome to represent the actual behavior of their application. Modern Chrome versions implement many (if not all) of the above ES2015 features, while JSC on iOS and Android do not. Therefore, something may work in the Chrome debugging environment but may not work properly when built for the device.

Thoughts?

**Comment by zackify** [comment]
I like your idea @skevy . I think it should be really clear on the getting started page that there are differences in the JS engine on android / iOS and that you should look into a polyfill if you're using any of the above ES-next features.

**Comment by vjeux** [comment]
@DmitrySoshnikov do you mind chiming in and figure out how to solve this problem for React Native as you probably did it for www?

**Comment by DmitrySoshnikov** [comment]
@skevy, we use `babel-plugin-symbol-member` for a while, and it works as intended.

> `Array.prototype[@@iterator]()` needs to exist

Do you guys use Babel for `for-of` or the native one? If Babel, then it even does an optimization for arrays and don't spawn an iterator, instead it's a simple `for`-loop in this case (if the instance is `Array.isArray(...)`).

If native `for-of` is available, then `Array.prototype[Symbol.iterator]` should also exist. Or if not, you can always implement one: it should reflect `values()` method:

```js
Array.prototype[Symbol.iterator] = function() {
  return this.values();
};
```

Note: in every place using the `babel-plugin-symbol-member` transform, you should use exactly `Symbol.iterator` (not `'@@iterator'`) in order to guarantee correct work in environments where `Symbol` is supported, and where it's not.

**Comment by skevy** [comment]
@DmitrySoshnikov it's late where I am, so I'll look into this more in the morning. Thanks so much for your response.

I will say, however, that at least in the Android JSC environment, `Array.prototype[Symbol.iterator]` does not exist...as after I add `babel-plugin-symbol-member` it transforms appropriately, but tells me that '@@iterator' doesn't exist on `Array.prototype`.

**Comment by skevy** [comment]
Just took a look at this deeper actually, and took a look at the Babel transform.

This is the case that breaks: https://astexplorer.net/#/WidiKglEp5

Notice how the `for..of` on the Array does as you say, and transforms to the for loop, but the `for..of` on Object.keys (which, on non-Symbol enabled platforms, such as our copy of JSC on Android, simply returns an Array) transforms to the iterator instance. Makes perfect sense of course, but that's what's causing our problem.

After inspecting the plugin though, if we ran in loose mode on that plugin, it would fix the issue and fall back to looping through like a normal array. So, that fixes `for..of`.

So would your recommendation be to then run in both loose mode on the `for..of` plugin and also include `babel-plugin-symbol-member`?

**Comment by DmitrySoshnikov** [comment]
@skevy, yeah, we use exactly loose mode, and seems it's more than enough in case of `for-of`. In this case it does `Array.isArray(...)`. However, you still may define non-enumerable `Symbol.iterator` on `Array.prototype` if needed.

**Comment by skevy** [comment]
@DmitrySoshnikov what would be the specific point of defining `Symbol.iterator` on `Array.prototype`, if we're using the loose mode version of `for..of`? The loose mode transform covers both the case where `Symbol` does exist on the platform, and when it does not. Is there another use case for `Symbol.iterator` on `Array` besides `for..of`?

**Comment by DmitrySoshnikov** [comment]
@skevy, I meant in case if you don't wanna use loose mode. Otherwise it should be fine (we don't have `Symbol.iterator` on `Array.prototype` internally).

**Comment by skevy** [comment]
@DmitrySoshnikov great. We'll go with using loose mode + documenting what we do and don't support.

@vjeux you good with this plan?

**Comment by vjeux** [comment]
Works for me, Dmitry is the pro here :p

**Comment by skevy** [comment]
Ok - I'm going to close this then in favor of adding loose mode in #5214 and then I'll open a new PR with doc changes.

**Comment by peterlazar1993** [comment]
I'm still facing this issue, using redux-saga on Android on react-native v0.20. This works in debug mode though.
screenshot_20160221-220844

**Comment by ms88privat** [comment]
:+1:

**Comment by lrettig** [comment]
I'm still getting this error in RN 0.20 on iOS 8.4 when trying to use `for...of`:

`ReferenceError: Can't find variable: Symbol`

**Comment by lrettig** [comment]
I upgraded to RN 0.20 and I'm now getting a different error. My code looks like this:

`for (let el of myArray) { ... }`

and it appears to transpile into this:

`for (var _iterator=myArray[typeof Symbol==='function'?Symbol.iterator:'@@iterator'](),_step; !(_iteratorNormalCompletion=(_step=_iterator.next()).done); _iteratorNormalCompletion=true) {
var el=_step.value;
... }`

which produces this error:

`TypeError: undefined is not a function (evaluating 'myArray[typeof Symbol==='function'?Symbol.iterator:'@@iterator']()')`

Indeed, my array doesn't contain functions, so I don't understand how this transpiled code is supposed to work. Does anyone know what's going on?

Thanks.

**Comment by skevy** [comment]
@lrettig in JS environments that support `Symbol`, the prototype of your Array would indeed contain a function: `myArray[Symbol.iterator]`. But in non JS envs, that Symbol.iterator is simply added to the Array prototype as `@@iterator`...so that's what this is checking.

I'm not sure why that's not working for you...are you using any custom babelrc or any type of custom config?

**Comment by lrettig** [comment]
There was indeed some junk in my .babelrc that must have caused this. I removed it, and it started working. Thanks!

**Comment by skevy** [comment]
So the reason, for what it's worth, that it wasn't working...is because your code was not being transpiled into the "for...of" loose mode code.

This is important for non-Symbol supporting JS platforms.

**Comment by lrettig** [comment]
I suspect the culprit was this entry in my .babelrc "plugins":
"transform-es2015-for-of" (probably from an earlier attempt to fix the same
problem).

**Comment by skevy** [comment]
Note that we have a preset now...called "babel-preset-react-native" that you can install and use in your babelrc, rather than copying the RN config or using the `extends` babelrc option.

This ensures that you're using all the proper transforms that RN requires.

**Comment by davodesign84** [comment]
related to this by any chance? https://github.com/gaearon/react-proxy/issues/50

**Comment by sdeleon28** [comment]
This missing polyfill causes this weird behavior with immutable https://github.com/facebook/immutable-js/issues/1474

Thought I'd leave it here for reference

**Pull request closed without merging**